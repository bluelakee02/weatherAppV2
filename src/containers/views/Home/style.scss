.responsive {
  min-width: 100%;
  min-height: 100%;
  background-position: center;
  background-size: cover;
  background-repeat: no-repeat;
  padding: 30px;
  border-radius: 5px;
}

.weatherCardSection {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  height: 500px;
}

.container {
  display: grid;
  grid-template-rows: 1fr 5fr;
  grid-template-columns: 1fr 1fr;
  grid-template-areas:
    "forecast image"
    "forecast image";
  justify-content: center;
  align-items: flex-start;
  flex-flow: column;
}

.subSection {
  display: flex;
  justify-content: flex-start;
  align-items: center;
  flex-wrap: wrap;
}

.mainSection {
  grid-area: forecast;
}

.searchSection {
  display: flex;
  flex-flow: column;
  justify-content: flex-start;
  align-items: flex-start;
  margin: 10px;
}

.imageSection {
  grid-area: image;
  width: 100%;
}

@media screen and (max-width: $breakpoint-desktop) {
  .subSection {
    justify-content: center;
    align-items: center;
  }

  .container {
    grid-template-rows: auto;
    grid-template-columns: 1fr;
    grid-template-areas:
    "forecast"
    "image";
  }

  .weatherCardSection {
    height: 100%;
  }

}
