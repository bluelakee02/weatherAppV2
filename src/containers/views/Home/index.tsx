import { RouteComponentProps } from '@reach/router';
import React, { useEffect, useState } from 'react';

import style from './style.scss';

import Clear from '@/assets/images/clear.webp';
import HeavyCloud from '@/assets/images/heavyCloud.webp';
import LightCloud from '@/assets/images/lightCloud.webp';
import Rain from '@/assets/images/rain.webp';
import Showers from '@/assets/images/showers.webp';
import Sleet from '@/assets/images/sleet.webp';
import Snow from '@/assets/images/snow.webp';
import ThunderStorm from '@/assets/images/thunderStorm.webp';
import Button from '@/components/Button';
import TodayWeatherCard, { Props } from '@/components/TodayWeatherCard';
import WeatherCard from '@/components/WeatherCard';

const getWeatherImage = (weatherState: string) => {
    switch (weatherState) {
        case 'Showers':
            return Showers;
        case 'Light Rain':
        case 'Heavy Rain':
        case 'Patchy rain possible':
            return Rain;
        case 'Heavy Cloud':
        case 'Cloudy':
        case 'Overcast':
            return HeavyCloud;
        case 'Sunny':
        case 'Clear':
            return Clear;
        case 'Thunderstorm':
        case 'Thundery outbreaks possible':
            return ThunderStorm;
        case 'Hail':
        case 'Sleet':
        case 'Fog':
        case 'Patchy freezing drizzle possible':
            return Sleet;
        case 'Snow':
        case 'Blowing snow':
        case 'Blizzard':
        case 'Patchy snow possible':
            return Snow;
        default:
            return LightCloud;
    }
};

const Home: React.FC<RouteComponentProps> = () => {
    const [locationQuery, setLocationQuery] = useState('');
    const [location, setLocation] = useState({ error: undefined, name: '' });
    const [forecast, setForecast] = useState([]);
    const [loading, setLoading] = useState(false);
    const [today, setToday] = useState<Props>();
    const { error, name } = location;

    const onSubmitLocation = async (coords = '') => {
        try {
            setLoading(true);
            const response = await fetch(`/getLocation?location=${coords ? coords : locationQuery}`);
            setLoading(false);
            const data = await response.json();
            const [locationData, today, ...forecastData] = data;
            setLocation(locationData);
            setToday(today[0]);
            setForecast(forecastData);
        } catch (err) {
            console.error(err);
        }
    };

    useEffect(() => {
        if (navigator.geolocation) {
            navigator.geolocation.getCurrentPosition(async position => {
                await onSubmitLocation(`${position.coords.latitude},${position.coords.longitude}`);
            });
        }
    }, []);

    const onKeyDown = async (event: React.KeyboardEvent<HTMLDivElement>) => {
        if (event.key === 'Enter') {
            event.preventDefault();
            event.stopPropagation();
            await onSubmitLocation();
        }
    };

    const { condition } = today || {};

    return (
        <div className={style.container}>
            <section className={style.mainSection}>
                <div className={style.subSection}>
                    <div className={style.searchSection}>
                        <input
                            type={'text'}
                            id={'location'}
                            name={'location'}
                            placeholder={'Type location'}
                            onChange={e => setLocationQuery(e.target.value)}
                            onKeyDown={onKeyDown}
                            disabled={loading}
                        />
                        <Button classname={style.button} label={'Search'} onClick={() => onSubmitLocation()} />
                        <div className={style.searchSection} style={{ height: '16px' }}>
                            {loading && <div>Loading ... </div>} {error && <div>Error: {error}</div>}
                        </div>
                    </div>
                </div>
                {condition && (
                    <div className={style.weatherCardSection}>
                        <div className={style.responsive} style={{ backgroundImage: `url(${getWeatherImage(condition || '')})` }}>
                            <h2 className={style.title}>{name} Today: </h2>
                            <TodayWeatherCard {...today} />
                        </div>
                    </div>
                )}
            </section>
            <section className={style.forecastSection}>
                <div className={style.subSection}>
                    {forecast.map((forecastData, idx) => (
                        <WeatherCard {...forecastData} key={idx} />
                    ))}
                </div>
            </section>
        </div>
    );
};

export default Home;
