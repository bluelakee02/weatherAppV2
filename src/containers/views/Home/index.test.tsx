import * as React from 'react';
import { shallow, render, mount } from 'enzyme';
import Home from './index';

describe('Home test', () => {
  it('Matches the snapshot', () => {
    expect(shallow(<Home />)).toMatchSnapshot();
  });

  it('Has proper class', () => {
    const wrapper = mount(<Home />);
    expect(wrapper.find(`.container`)).toHaveLength(1);
  });

  it('Has proper nodes', async () => {
    let wrapper = mount(<Home />);

    expect(wrapper.find('h1').length).toBe(0);
    expect(wrapper.find('img').length).toBe(0);
    expect(wrapper.find('input').length).toBe(1);
    expect(wrapper.find('button').length).toBe(1);

  });

  it('Renders static HTML', () => {
    expect(render(<Home />).text()).toContain('Search');
  });
});
