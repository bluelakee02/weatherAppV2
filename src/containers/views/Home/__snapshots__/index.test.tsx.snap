// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home test Matches the snapshot 1`] = `
<div
  className="container"
>
  <section
    className="mainSection"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      className="subSection"
    >
      <div
        className="searchSection"
      >
        <input
          disabled={false}
          id="location"
          name="location"
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Type location"
          type="text"
        />
        <Button
          label="Search"
          onClick={[Function]}
        />
      </div>
      <div
        className="searchSection"
        style={
          Object {
            "height": "75px",
          }
        }
      />
    </div>
    <div
      className="subSection"
    />
  </section>
</div>
`;
